paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,497 @@
+  }
+
+  virtual mkldnn::memory CreateDstMemory(
+      const mkldnn::inner_product_forward::primitive_desc& fc_prim_desc,
+      const ExecutionContext& ctx, Tensor* output,
+      bool force_fp32_output = false) {
+    auto dst_prim_desc = fc_prim_desc.dst_primitive_desc();
+    auto buffer_size = dst_prim_desc.get_size();
+
+    /* ADD: force fp32 output */
+    if (force_fp32_output) {
+      float* output_data = output->mutable_data<float>(
+          ctx.GetPlace(), ::paddle::memory::Allocator::kDefault, buffer_size);
+      output->set_format((memory::format)dst_prim_desc.desc().data.format);
+      return memory(dst_prim_desc, to_void_cast<float>(output_data));
+    }
+
+    int8_t* output_data = output->mutable_data<int8_t>(
+        ctx.GetPlace(), ::paddle::memory::Allocator::kDefault, buffer_size);
+    output->set_format((memory::format)dst_prim_desc.desc().data.format);
+    return memory(dst_prim_desc, to_void_cast<int8_t>(output_data));
+  }
+};
+
+template <typename T, typename K>
+std::shared_ptr<MulPrimitiveFactory<T, K>> GetPrimitiveFactory(
+    const MKLDNNDeviceContext& dev_ctx, const ExecutionContext& ctx,
+    const Tensor* input, const Tensor* weights,
+    const mkldnn::engine& mkldnn_engine) {
+  const std::string key = GetHash(input, weights, ctx.op().Output("Out"));
+
+  auto prim_creator =
+      std::static_pointer_cast<MulPrimitiveFactory<T, K>>(dev_ctx.GetBlob(key));
+  if (prim_creator == nullptr) {
+    bool enable_quant =
+        std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value;
+    if (enable_quant) {
+      prim_creator =
+          std::make_shared<QuantMulPrimitiveFactory<T, K>>(mkldnn_engine);
+    } else {
+      prim_creator = std::make_shared<MulPrimitiveFactory<T, K>>(mkldnn_engine);
+    }
+    dev_ctx.SetBlob(key, prim_creator);
+  }
+
+  return prim_creator;
+}
+
+template <typename T, typename K>
+class MulMKLDNNKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    PADDLE_ENFORCE(paddle::platform::is_cpu_place(ctx.GetPlace()),
+                   "It must use CPUPlace.");
+
+    auto& dev_ctx =
+        ctx.template device_context<paddle::platform::MKLDNNDeviceContext>();
+    const auto& mkldnn_engine = dev_ctx.GetEngine();
+
+    const Tensor* x = ctx.Input<Tensor>("X");
+    const Tensor* y = ctx.Input<Tensor>("Y");
+    Tensor* z = ctx.Output<Tensor>("Out");
+
+    auto prim_creator =
+        GetPrimitiveFactory<T, K>(dev_ctx, ctx, x, y, mkldnn_engine);
+    auto fc = prim_creator->CreateMulPrimitive(x, y, z, ctx);
+
+    stream(stream::kind::eager).submit({fc}).wait();
+
+    auto z_dim = z->dims();
+    z->set_layout(DataLayout::kMKLDNN);
+    z->set_format(z_dim.size() != 2
+                      ? platform::MKLDNNFormatForSize(z_dim.size(), x->format())
+                      : z->format());
+
+    if (z_dim.size() != 2) {
+      z->Resize(z_dim);
+    }
```

Comment:
Is it possible that output is not 2 dimensional?

Reply:
no, the output should always be 2 dims, will remove it